### PR TITLE
LibWeb+LibGfx: Add basic SVG `<mask>` support 

### PR DESCRIPTION
**Pull request by MacDue**
This PR adds basic support for the SVG `<mask>` element (and attribute). It leaves many TODOs for now, and only implements the default case:

- [maskContentUnits](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/maskContentUnits) = userSpaceOnUse
- [maskUnits](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/maskUnits) = objectBoundingBox
- mask-type = luminance 
- Default masking area (x: -10%, y: -10%, width: 120%, height: 120%)

This is still enough for a decent number of real world examples.

---

Example SVG masking (from MDN)

```html
<svg width="200" viewBox="-10 -10 120 120">
  <rect x="-10" y="-10" width="120" height="120" fill="blue" />
  <mask id="myMask">
    <!-- Everything under a white pixel will be visible -->
    <rect x="0" y="0" width="100" height="100" fill="white" />
    <!-- Everything under a black pixel will be invisible -->
    <path
      d="M10,35 A20,20,0,0,1,50,35 A20,20,0,0,1,90,35 Q90,65,50,95 Q10,65,10,35 Z"
      fill="black" />
  </mask>
  <polygon points="-10,110 110,110 110,-10" fill="orange" />
  <!-- with this mask applied, we "punch" a heart shape hole into the circle -->
  <circle cx="50" cy="50" r="50" fill="purple" mask="url(#myMask)" />
</svg>
```
Results in:
image

--- 

Progression on GitHub -- the sparklines now work. SVG masking is used to mask out the line from a `<rect>` filled with a gradient:

Screenshot from 2023-09-10 14-46-36

### CODE REVIEW

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Creating layout and paintable nodes for the `<mask>` seems odd to me, but I'm not familiar enough with how our SVG implementation works to say that it's actually wrong. Just seems strange for something to have them when it's not actually visible in itself.

**Comment by MacDue** [comment]
> Creating layout and paintable nodes for the <mask> seems odd to me
How else would it be possible to paint them? The contents of a mask need to be laid out, and they also need to be painted, just to an off-screen buffer rather than directly to the screen.

**Comment by AtkinsSJ** [comment]
> > Creating layout and paintable nodes for the  seems odd to me
> 
> How else would it be possible to paint them? The contents of a mask needs to be laid out, and it also needs to be painted, just to an off-screen buffer rather than directly to the screen.

But you don't actually know what size to paint them until later... I think mostly this is me not having a good understanding. :sweat_smile:

**Comment by MacDue** [comment]
> > > Creating layout and paintable nodes for the  seems odd to me
> > 
> > 
> > How else would it be possible to paint them? The contents of a mask needs to be laid out, and it also needs to be painted, just to an off-screen buffer rather than directly to the screen.
> 
> But you don't actually know what size to paint them until later... I think mostly this is me not having a good understanding. 😅

The layout of the masks implemented here do not depend on the element they are masking. The mask is positioned within the SVG like any other element, it just happens to be composited in a different way. 

Even if they did, I think the sensible solution to that is to apply an affine transformation while painting the mask. Computing layout + paintables during painting, or duplicating the mask for each target element seems silly :)

File: Userland/Libraries/LibWeb/Painting/PaintableBox.h [hunk 1]
```diff
@@ -25,7 +25,7 @@ class PaintableBox : public Paintable {
 
     virtual void paint(PaintContext&, PaintPhase) const override;
 
-    bool is_visible() const { return layout_box().is_visible(); }
+    virtual bool is_visible(Optional<PaintContext const&> = {}) const { return layout_box().is_visible(); }
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Painting/PaintableBox.h]: Not sure why this function needs to have two modes, can we just always pass `PaintContext const&` to keep it simple? :)

File: Userland/Libraries/LibWeb/Painting/SVGPaintable.cpp [hunk 2]
```diff
@@ -31,4 +33,48 @@ CSSPixelRect SVGPaintable::compute_absolute_rect() const
     return PaintableBox::compute_absolute_rect();
 }
 
+Optional<CSSPixelRect> SVGPaintable::get_masking_area() const
+{
+    if (!is<SVG::SVGGraphicsElement>(dom_node()))
+        return {};
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Painting/SVGPaintable.cpp:39]: Would it make sense to put this function in `SVGGraphicsPaintable` instead?

File: Userland/Libraries/LibWeb/Painting/SVGPaintable.cpp [hunk 3]
```diff
@@ -31,4 +33,48 @@ CSSPixelRect SVGPaintable::compute_absolute_rect() const
...
+    auto const& graphics_element = static_cast<SVG::SVGGraphicsElement const&>(*dom_node());
+    if (auto mask = graphics_element.mask())
+        return mask->resolve_masking_area(absolute_border_box_rect());
+    return {};
+}
+
+void SVGPaintable::apply_mask(PaintContext& context, Gfx::Bitmap& target, CSSPixelRect const& masking_area) const
+{
+    if (!is<SVG::SVGGraphicsElement>(dom_node()))
+        return;
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Painting/SVGPaintable.cpp:49]: Would it make sense to put this function in `SVGGraphicsPaintable` instead?

File: Userland/Libraries/LibWeb/Painting/SVGPaintable.cpp (hunk 2 above)
Comment (MacDue) [reply Userland/Libraries/LibWeb/Painting/SVGPaintable.cpp]: Yep :+1:

File: Userland/Libraries/LibWeb/Painting/PaintableBox.h (hunk 1 above)
Comment (MacDue) [reply Userland/Libraries/LibWeb/Painting/PaintableBox.h]: Previously, it needed to be optional as `is_visible()` was used from hit testing too. I've now removed this change and instead added `Paintable::forms_unconnected_subtree()`, which is true for SVG masks. That means SVG masks won't be connected to the (main) paint tree, so can't be painted (unless referenced by another element's `mask` attribute). I think that's a little nicer of a solution.

Comment (awesomekling) [reply Userland/Libraries/LibWeb/Painting/PaintableBox.h]: That does seem nicer!

**Pull request merged**